### PR TITLE
poc conda hosted repository

### PR DESCRIPTION
**Pull request by jaroslawk**
This is the initial implementation of conda-hosted repository.

Features working:
* allow of uploading conda packages using PUT call
* recreation of repodata.json for each upload 
* fetching uploaded packages 

Missing TODO:
* recreation of repodata.json need to be started on event

cc @DarthHater @bhamail

### CODE REVIEW

**Comment by jaroslawk** [comment]
@DarthHater Great. Thank you.

I got really cryptic situation doing the redeploy of the artifact - I would appreciate any help with this. WIll generate and issue with details.

**Comment by jaroslawk** [comment]
@DarthHater how is the review progressing? Is there anything I can help you with?

**Comment by DarthHater** [comment]
Hah, you actually fixed one of the issues I was looking in to (the redeploy). @bhamail you got any cycles to take a gander here?

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedFacetImpl.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+        if (asset == null) {
+            return false;
+        }
+        tx.deleteAsset(asset);
+        return true;
+    }
+
+
+    protected Content doPutContent(final String path, final TempBlob tempBlob, final Payload payload) throws IOException {
+
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedFacetImpl.java]: TRIVIAL: Newline

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedFacetImpl.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    private RepoData createRepoData(String arch) {
+        RepoData repoData = new RepoData();
+        repoData.getInfo().setSubdir(arch);
+        return repoData;
+    }
+
+    @TransactionalStoreMetadata
+    private Asset getOrCreateAsset(final CondaPath condaPath, final Repository repository) {
+
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedFacetImpl.java]: TRIVIAL: Newline

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedRecipe.groovy [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        repository.attach(configure(viewFacet.get()))
+        repository.attach(httpClientFacet.get())
+        repository.attach(componentMaintenanceFacet.get())
+        repository.attach(storageFacet.get())
+        repository.attach(hostedFacet.get())
+        repository.attach(searchFacet.get())
+        repository.attach(attributesFacet.get())
+    }
+
+
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedRecipe.groovy]: TRIVIAL: Superflous newline

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedRecipe.groovy [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    static Matcher fetchMetaDataMatcher = and(new ActionMatcher(GET, HEAD), condaMetaDataMather)
+    static Matcher fetchCondaFileMatcher = and(new ActionMatcher(GET, HEAD), condaPathMatcher)
+    static Matcher uploadCondaFileMatcher = and(new ActionMatcher(PUT), condaPathMatcher)
+    static Matcher deleteCondaFileMatcher = and(new ActionMatcher(DELETE), condaPathMatcher)
+
+    static String removeLeadingSlash(String path) {
+        path.length() > 1 && path[0] == '/' ? path.substring(1) : path
+    }
+
+    final Handler handler = new Handler() {
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedRecipe.groovy]: It's interesting how you've implemented this. I've tended in the past to do something more akin to: https://github.com/sonatype-nexus-community/nexus-repository-helm/blob/master/src/main/java/org/sonatype/repository/helm/internal/hosted/HostedHandlers.java

I would suggest moving them out of the recipe and into their own class, etc...

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedRecipe.groovy [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    static Matcher condaMetaDataMather = {
+        Context context ->
+            def path = context.getRequest().getPath()
+            log.warn("Searching condaMetaDataMather " + path)
+            String candidate = path.substring(path.lastIndexOf('/'))
+            CondaPathUtils.CONDA_META_DATA.contains(candidate)
+    }
+
+    /**
+     * Conda path matcher - verifies path has conda alike structure
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaHostedRecipe.groovy]: Typo: I think it should be `verifies path has conda like structure`

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaPath.java [hunk 6]
```diff
@@ -0,0 +1,264 @@
+package org.sonatype.nexus.plugins.conda.internal.hosted;
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaPath.java]: License missing

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaWritePolicySelector.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package org.sonatype.nexus.plugins.conda.internal.hosted;
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaWritePolicySelector.java]: Missing license

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaUploadHandler.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
+package org.sonatype.nexus.plugins.conda.internal.hosted;
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/CondaUploadHandler.java]: Missing license

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/metadata/MetaData.groovy [hunk 9]
```diff
@@ -0,0 +1,109 @@
+package org.sonatype.nexus.plugins.conda.internal.hosted.metadata
+
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/metadata/MetaData.groovy]: TRIVIAL: Superfluous newline

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/metadata/MetaData.groovy [hunk 10]
```diff
@@ -0,0 +1,109 @@
+package org.sonatype.nexus.plugins.conda.internal.hosted.metadata
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/metadata/MetaData.groovy]: Missing license

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/metadata/MetaData.groovy [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+    static RepoData asRepoData(String json) {
+        def cleanJson = json.replaceAll('packages.conda', 'packages_conda')
+        new JsonSlurper().parseText(cleanJson) as RepoData
+    }
+
+    static ChannelData asChannelData(String json) {
+        new JsonSlurper().parseText(json) as ChannelData
+    }
+
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/hosted/metadata/MetaData.groovy]: TRIVIAL: Superfluous newline

File: src/main/java/org/sonatype/nexus/plugins/conda/internal/proxy/CondaProxyFacetImpl.java [hunk 12]
```diff
@@ -65,7 +66,7 @@ public CondaProxyFacetImpl(final CondaPathUtils condaPathUtils,
     this.condaDataAccess = checkNotNull(condaDataAccess);
   }
 
-  // HACK: Workaround for known CGLIB issue, forces an Import-Package for org.sonatype.nexus.repository.config
+  // HACK: Workaround for known CGLIB issue, forces an Import-PackageIndex for org.sonatype.nexus.repository.config
```
Comment (DarthHater) [inline src/main/java/org/sonatype/nexus/plugins/conda/internal/proxy/CondaProxyFacetImpl.java]: Question: Was this change on purpose?

File: src/main/resources/static/rapture/NX/conda/view/repository/recipe/CondaHosted.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    var me = this;
+
+    me.items = [
+      { xtype: 'nx-coreui-repository-storage-facet'},
+      { xtype: 'nx-coreui-repository-storage-hosted-facet'}
+    ];
+
+    me.callParent();
+  }
+});
```
Comment (DarthHater) [inline src/main/resources/static/rapture/NX/conda/view/repository/recipe/CondaHosted.js]: TRIVIAL: Newline at EOF

File: src/test/java/org/sonatype/nexus/plugins/conda/internal/hosted/MetaDataTest.groovy [hunk 14]
```diff
@@ -0,0 +1,164 @@
+package org.sonatype.nexus.plugins.conda.internal.hosted
```
Comment (DarthHater) [inline src/test/java/org/sonatype/nexus/plugins/conda/internal/hosted/MetaDataTest.groovy]: Missing license

**Comment by jaroslawk** [comment]
@DarthHater I have fixed the cosmetic (new liners/licence) and extracted Handlers. 
- Is there any auto-format style template could be used? 
- what do you think of using Spock as unit testing framework?

**Comment by jaroslawk** [comment]
for some reason, CICD is not kicking in... can this be fixed?

**Comment by bhamail** [comment]
@jaroslawk This is way cool! I will review the code as well, but wanted to first focus on your question about CI. I think it didn't run because I had the "Build forked pull requests" turned off. That is turned on now, but I'm not sure CI will automagically wake-up to build an existing branch. As a test, could you please push a simple change to your branch, and we'll see if that triggers a build?

As a heads-up, I've been working on a refactoring of this project to include Integration Tests here: PR #23 
Don't let that work deter you, but wanted to point it out. If you felt like rebasing onto that work, that would be cool, but don't feel obligated to do so - you beat me to the PR fair and square! :)

File: src/test/java/org/sonatype/nexus/plugins/conda/internal/hosted/MetaDataTest.groovy [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+  "name": "mlflow",
+  "noarch": "python",
+  "platform": null,
+  "subdir": "noarch",
+  "timestamp": 1566283016272,
+  "version": "1.2.0"
+}'''
+
+    void test_read_index_json_from_tar_bz2_archive() {
+        def input = MetaDataTest.getClass().getResourceAsStream("/sample_package-1.2.0-py_0.tar.bz2")
```
Comment (bhamail) [inline src/test/java/org/sonatype/nexus/plugins/conda/internal/hosted/MetaDataTest.groovy]: Is a resource file missing from your commit?

I don't see that file anywhere, and I get the following error when running this test:

```
[ERROR] Errors: 
[ERROR]   MetaDataTest.test_read_index_json_from_tar_bz2_archive:145 » NullPointer
```

Comment (jaroslawk) [reply src/test/java/org/sonatype/nexus/plugins/conda/internal/hosted/MetaDataTest.groovy]: correct it was missing resource - fixed now

**Comment by jaroslawk** [comment]
@bhamail 
- about the Integration test PR - my is touching one package only. Let's review it and merge it first ok? It will be fairly easy to merge my changes to yours.

- CICD I see the checks are kicked in but nothing happens. Someone with circleci access needs to take a look...

**Comment by bhamail** [comment]
@jaroslawk 

* Waiting to add ITs after this is merged is totally fine with me.

* Maybe it was just a delay, but the required parts of the CI appear to have run and succeeded:

  CIRequiredRan

  You can ignore the other non-required "workflow" builds.

**Comment by jaroslawk** [comment]
@bhamail I was hoping to see unit test results :) 
Let's review this PR and get it merged then.

**Comment by bhamail** [comment]
@jaroslawk I apologize for not seeing this forest for the trees sooner: 
The Conda format is now part of Nexus Repository Manager.
The best approach is to create a new PR for your Hosted Conda format against the [nexus-public](https://github.com/sonatype/nexus-public) repository.
Specifically, take a look at the source for the [Conda format](https://github.com/sonatype/nexus-public/tree/master/plugins/nexus-repository-conda).

I should probably close down this "old" Conda community format repo to avoid a similar problem like this in the future. We've been using this project as a test bed for a number of experiments, but if it cause confusion, it should probably be retired. At the very least, I will add a pointer to the top level README.md pointing the to official format.

**Comment by jaroslawk** [comment]
@bhamail this is really unfortunate. 

Please state this clearly in repo readme pointing to the proper one
Closing PR.

**Pull request closed without merging**